### DISCUSSION TITLE
MYSQL round up 23:59:59.999 to the next day on insert

### DISCUSSION THREAD

**Discussion by itxtoledo** [mikro-orm/mikro-orm#4633] [Ideas]
**Describe the bug**
Mysql rounds from a date to the nearest date if the time ends in 999 for example: 2023-08-17 23:59:59.999 will be rounded to 2023-08-18 00:00:00

**To Reproduce**
1. try to insert date using "endOfDay" function from date-fns that will show hour as: 23:59:59.999
2. look at your mysql table, the date added is the next day at 00:00:00
3. how to solve: subtract 2ms from date like these aswers explains:

"EndDate + 23:59:59.999 would get rounded up to the next day anyway. It would need to be EndDate + 23:59:59.997 to avoid that." 
Ref: https://stackoverflow.com/questions/7691742/add-time-235959-999-to-end-date-for-between

"datetime values are rounded to increments of .000, .003, or .007 seconds"
Ref: https://stackoverflow.com/questions/25760075/why-cast20140904-235959-999-as-datetime-equals-2014-09-05-000000-000

**Expected behavior**
I expected the ORM to detect that I am using the mysql driver and automatically subtract the 2 ms for the end of the date to be 997.

**Additional context**
I'm using "endOfDay" function from date-fns.

**Versions**

| Dependency | Version |
| - | - |
| node | v18.17.0 |
| typescript | 5.1.6 |
| mikro-orm | 5.7.12 |
| your-driver | "@mikro-orm/mysql": "^5.7.12" |

**Comment by B4nan** [comment]
So you are storing the datetime without milliseconds, do I get that right?

I don't think this is something the ORM should be fixing anyhow, moreover, the rounding seems to be a correct behavior to me. Do you know some tools that would be dealing with this? 

In v6 you will be able to have a custom type for this, I am working on a refactoring that will alter how the dates are processed and it will allow full control (mapping the raw strings from the database and back, in any way you want). You could probably already define a custom type and just truncate the values yourself, but in v6 you will be in control of the value provided directly by the database, in v5 things are automatically converted to Date objects before the value reaches the custom type.

**Reply by itxtoledo** [reply]
In sequelize which is the ORM I used before knowing mikroorm it seems to me that it removes the milliseconds before inserting since mysql does not store this data, see the print as it is stored in mysql

![image](https://github.com/mikro-orm/mikro-orm/assets/21026675/e635e0b0-dec3-4f58-b40e-333b9625c437)

**Reply by itxtoledo** [reply]
it's not a big problem, I can define the ms as 0 in my code, it took me a while to understand where the problem was

**Reply by B4nan** [reply]
TBH it feels pretty weird that `.997` rounds differently than `.999` in the first place.

**Reply by itxtoledo** [reply]
In sequelize it seems that the milliseconds are removed, in mikroorm it is not so mysql rounds up or down since it does not store milliseconds, only hours, minutes and seconds

**Reply by B4nan** [reply]
FYI MySQL can store milliseconds just fine, it's just about defaults for the precision.
